ufl/algorithms/transformer.py

```diff
@@ -216,11 +216,35 @@ def variable(self, o):
         return self.visit(o.ufl_operands[0])
 
 
-def apply_transformer(e, transformer, integral_type=None):
+class FormArgumentStripper(CopyTransformer):
+    def __init__(self):
+        CopyTransformer.__init__(self)
+        self.arg_map = {}
+
+    def terminal(self, o):
+        new_o = o.ufl_strip_data()
+        self.arg_map[new_o] = o
+        return new_o
+
+
+class FormArgumentReplacer(CopyTransformer):
+    def __init__(self, arg_map):
+        CopyTransformer.__init__(self)
+        self.arg_map = arg_map
+
+    def terminal(self, o):
+        try:
+            return self.arg_map[o]
+        except KeyError:
+            raise ValueError(f"FormArgument {o} is not present in the mapping")
```

Comment:
You weren't to know this, but the right thing to inherit from is not this `Transformer` but rather the `MultiFunction` in `ufl.corealg.multifunction`.

This one can almost certainly just be implemented with `ufl.replace`.

The stripper could too I think as:

```
ufl.replace(form, dict(zip(o, o.ufl_strip_data()) for o in form.terminals()))
```

(Pseudo-code).